File: learn/hashing/double-hashing/double_hashing.go


type DoubleHashItem struct {
	key     int
	value   int
	deleted bool
}

type DoubleHashing struct {
	capacity int
	size     int
	table    []*DoubleHashItem
	prime    int
}

func NewDoubleHashing(cap, prime int) *DoubleHashing {
	return &DoubleHashing{
		capacity: cap,
		size:     0,
		table:    make([]*DoubleHashItem, cap),
		prime:    prime,
	}
}

func (dh *DoubleHashing) hash1(key int) int {
	return (key%dh.capacity + dh.capacity) % dh.capacity
}

func (dh *DoubleHashing) hash2(key int) int {
	return dh.prime - ((key%dh.prime + dh.prime) % dh.prime)
}

func (dh *DoubleHashing) Insert(key, value int) {
	if dh.size == dh.capacity {
		return
	}
	index := dh.hash1(key)
	step := dh.hash2(key)

	for dh.table[index] != nil && !dh.table[index].deleted && dh.table[index].key != key {
		index = (index + step) % dh.capacity
	}

	if dh.table[index] == nil || dh.table[index].deleted {
		dh.size++
	}
	dh.table[index] = &DoubleHashItem{key: key, value: value, deleted: false}
}

func (dh *DoubleHashing) Search(key int) (int, bool) {
	index := dh.hash1(key)
	step := dh.hash2(key)
	startIndex := index

	for dh.table[index] != nil {
		if !dh.table[index].deleted && dh.table[index].key == key {
			return dh.table[index].value, true
		}
		index = (index + step) % dh.capacity
		if index == startIndex {
			break
		}
	}
	return -1, false
}

func (dh *DoubleHashing) Remove(key int) {
	index := dh.hash1(key)
	step := dh.hash2(key)
	startIndex := index

	for dh.table[index] != nil {
		if !dh.table[index].deleted && dh.table[index].key == key {
			dh.table[index].deleted = true
			dh.size--
			return
		}
		index = (index + step) % dh.capacity
		if index == startIndex {
			break
		}
	}
}

File: learn/hashing/double-hashing/double_hashing.rs
#[derive(Clone)]
pub struct DoubleHashItem {
    pub key: i32,
    pub value: i32,
    pub deleted: bool,
}

pub struct DoubleHashing {
    capacity: usize,
    size: usize,
    table: Vec<Option<DoubleHashItem>>,
    prime: usize,
}

impl DoubleHashing {
    pub fn new(capacity: usize, prime: usize) -> Self {
        DoubleHashing {
            capacity,
            size: 0,
            table: vec![None; capacity],
            prime,
        }
    }

    fn hash1(&self, key: i32) -> usize {
        key.rem_euclid(self.capacity as i32) as usize
    }

    fn hash2(&self, key: i32) -> usize {
        self.prime - (key.rem_euclid(self.prime as i32) as usize)
    }

    pub fn insert(&mut self, key: i32, value: i32) {
        if self.size == self.capacity {
            return;
        }
        let mut index = self.hash1(key);
        let step = self.hash2(key);

        while let Some(ref item) = self.table[index] {
            if !item.deleted && item.key == key {
                break;
            }
            if item.deleted {
                break;
            }
            index = (index + step) % self.capacity;
        }

        if self.table[index].is_none() || self.table[index].as_ref().unwrap().deleted {
            self.size += 1;
        }
        self.table[index] = Some(DoubleHashItem {
            key,
            value,
            deleted: false,
        });
    }

    pub fn search(&self, key: i32) -> Option<i32> {
        let mut index = self.hash1(key);
        let step = self.hash2(key);
        let start_index = index;

        while let Some(ref item) = self.table[index] {
            if !item.deleted && item.key == key {
                return Some(item.value);
            }
            index = (index + step) % self.capacity;
            if index == start_index {
                break;
            }
        }
        None
    }

    pub fn remove(&mut self, key: i32) {
        let mut index = self.hash1(key);
        let step = self.hash2(key);
        let start_index = index;

        while let Some(ref mut item) = self.table[index] {
            if !item.deleted && item.key == key {
                item.deleted = true;
                self.size -= 1;
                return;
            }
            index = (index + step) % self.capacity;
            if index == start_index {
                break;
            }
        }
    }
}
